src/backend/executor/exchange_seq_scan_executor.cpp

```diff
@@ -0,0 +1,182 @@
+  const planner::SeqScanPlan &node = GetPlanNode<planner::SeqScanPlan>();
+  predicate_ = node.GetPredicate();
+  column_ids_ = std::move(node.GetColumnIds());
+
+  target_table_ = node.GetTable();
+  if (target_table_ != nullptr) {
+    tile_group_number_ = target_table_->GetTileGroupCount();
+    if (column_ids_.empty()) {
+      column_ids_.resize(target_table_->GetSchema()->GetColumnCount());
+      std::iota(column_ids_.begin(), column_ids_.end(), 0);
+    }
+  }
+
+  return true;
+}
+
+/*
+ * This method is not supposed to be called by multiple threads, even though
+ * we are implementing a parallelized executor.
+ */
+bool ExchangeSeqScanExecutor::DExecute() {
+  if (!done_) {
+    // Scanning over a logical tile.
+    if (children_.size() == 1) {
+      LOG_TRACE("Exchange Sequential Scan Executor : 1 child ");
+
+      assert(target_table_ == nullptr);
+      assert(column_ids_.size() == 0);
+      // In this case, just use one thread to pull data
+      while (children_[0]->Execute()) {
+        std::unique_ptr<LogicalTile> tile(children_[0]->GetOutput());
+        if (predicate_ != nullptr) {
+          // Invalidate tuples that don't satisfy the predicate.
+          for (oid_t tuple_id : *tile) {
+            expression::ContainerTuple<LogicalTile> tuple(tile.get(), tuple_id);
+            if (predicate_->Evaluate(&tuple, nullptr, executor_context_)
+                    .IsFalse()) {
+              tile->RemoveVisibility(tuple_id);
+            }
+          }
+        }
+        if (0 == tile->GetTupleCount()) {  // Avoid returning empty tiles
+          continue;
+        }
+        result_.emplace(tile.release());
+      }
+    }
+    // Scanning a table
+    else if (children_.size() == 0) {
+      LOG_TRACE("Exchange Sequential Scan Executor : 0 child ");
+
+      assert(target_table_ != nullptr);
+      assert(target_table_->GetTileGroupCount() == tile_group_number_);
+      assert(column_ids_.size() > 0);
+
+      // In this case, wrap the scanning of each tile group into an individual
+      // task
+      // Break the whole execution into multiple tasks
+      for (oid_t no = 0; no < tile_group_number_; ++no) {
+        ThreadManager::GetInstance().AddTask(
+            std::bind(&ExchangeSeqScanExecutor::ScanOneTileGroup, this, no,
+                      concurrency::current_txn));
+      }
+      {
+        // Wait for all tasks to be done
+        std::unique_lock<std::mutex> lock(result_lock_);
+        while (finished_number_ < tile_group_number_) cv_.wait(lock);
+      }
+    }
+    done_ = true;
+  }
+  // Having finished the entire scan
+  // Return results
+  assert(done_);
+  assert(finished_number_ == tile_group_number_);
+  if (result_.empty()) return false;
+  SetOutput(result_.front().release());
+  result_.pop();
+  return true;
+}
+
+/*
+ * Do a table scan on one tile group
+ */
+void ExchangeSeqScanExecutor::ScanOneTileGroup(
+    const oid_t no, concurrency::Transaction * /*unused*/) {
+  /*
```

Comment:
Thanks.
The transaction manager assumes that one transaction only uses one thread. Therefore, we thought all its methods were not thread-safe. But now I see that IsVisible function is actually a read-only function, which can be called by multiple threads. Therefore, we will add IsVisible back here. However, because of the reason in comment, the PerformRead functions cannot be used here.